packages/firestore/src/core/sync_engine.ts

```diff
@@ -490,6 +493,11 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
       // connection is disabled.
       await this.remoteStore.fillWritePipeline();
     } else if (batchState === 'acknowledged' || batchState === 'rejected') {
+      // If we receive a notification of an `acknowledged` or `rejected` batch
+      // via Web Storage, we are either already secondary or another tab has
+      // taken the primary lease.
```

Comment:
I do think this is racey, but it is racey both ways: If a client comes back from the dead, and thinks it is primary, it might take a while for it to catch up on Local Storage notifications before it realizes it is no longer primary. 

If I remove this line, I get additional query events about half the time and the tests become flakey. It loos like this is the only side effect, and if two primary tabs were to run this code in this function at the same time, the world would still be a happy place.

My argument for keeping it is another though: If a tab becomes primary even though it has not caught up on LocalStorage notifications, it seems much safer to go back into secondary mode and finish these notifications in the state it was supposed to be in until the client has caught up.

Reply:
I'm not sure I follow / agree-with your argument.  I don't think we can deterministically "finish these notifications in the state it was supposed to be in".

At the end of the day it's possible for two tabs to send the same write to the backend and broadcast LocalStorage notifications for it.  E.g. Tab A is primary, performs a write, gets the response, and gets paused / throttled right before sending the LocalStorage notification.  Tab B takes over as primary, re-sends the write, and processes the response (sending a LocalStorage notification). Tab A resumes and its LocalStorage notification (for the same write) now goes out.

I think this means that handling write acknowledgement notifications must be idempotent.

But I also don't understand where the extra query events you mentioned are coming from...  It may be worth talking through this or something.